fossas/fossa-cli
# Ane 891 dynamic go tactic

# Overview

This is an initial implementation of a new experimental Go strategy based on `go list -json -deps all`. 

This PR does not implement everything that needs to be done for ANE-891. But I wanted to get some review early. ACs that still need to be done in this or a future PR are:

- [ ] Exclude test dependencies
- [ ] Handle pseudo-versions
- [ ] Improve efficiency - This isn't a problem that I've seen so far, but there is a little low-hanging fruit in my implementation IMO. 

I've added some minimal documentation for this tactic, there is going to be another ticket to write something more extensive. I will update the changelog if reviewers think that this is ok to incorporate into a release.

Also, the flag itself is a bit long and unwieldy (`experimental-package-centric-go`) and also talks about implementation details, which I think is inappropriate. If anyone has other ideas I'd like to hear them.

## Acceptance criteria

   * There should be a new dynamic tactic for Go using `go list -json -deps all` available.
   * The new tactic is under an experimental flag and not be the default.
   * Should handle both local and package replacements.
   * Should Label deep and direct dependencies.
   * Should detect path dependencies and not include them or their child deps in the graph. 
   * Handle the "C" special package. 

## Testing plan

1. Build the cli on this branch.
2. Run `fossa analyze --experimental-package-centric-go <dir>` on a Go project.
3. Look over the results to make sure they make sense. 

Repos that demonstrate the functionality of this strategy:

https://github.com/oss-review-toolkit/ort/tree/7725f9285eb6d6671937384d0f2a7896cfeb3332/analyzer/src/funTest/assets/projects/synthetic/gomod - Shows removing modules with unused packages from graph. 
https://github.com/ipfs/kubo - Shows ignoring the special "C" package. The current strategy will also do this.
[local_path_deps.tar.gz](https://github.com/fossas/fossa-cli/files/11081554/local_path_deps.tar.gz) this demonstrates pruning out local path deps. You can see the difference by running it with the experimental flag and then without it.
https://github.com/kubernetes/kubernetes - (I found a bug with scanning this, I will update when I fix it)
https://github.com/coredns/coredns 


## Risks
Any major risks I can think of are inherent to trying to replace our dynamic Go tactic and not specific to this PR.

## References

- [ANE-891](https://fossa.atlassian.net/browse/ANE-891)

## Checklist



[ANE-891]: https://fossa.atlassian.net/browse/ANE-891?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
